File: ios/CallPilot/Content/CallHistoryModel.swift
import Combine
import Foundation

enum CallHistorySyncStatus: Equatable {
    case idle
    case loading
    case live
    case stale
    case offline
}

enum CallHistoryCopy {
    static var payloadTooLarge: String { L10n.text("calls.error.payload_too_large") }
    static var unavailable: String { L10n.text("calls.error.unavailable") }
    static var edgeOffline: String { L10n.text("calls.error.edge_offline") }
    static var featureDisabled: String { L10n.text("calls.error.feature_disabled") }
    static var unauthorized: String { L10n.text("content.error.unauthorized") }
}

enum CallSummaryPresentation: Equatable {
    case hidden
    case pending
    case ready(CallSummary?)
    case failed(CallSummary?)

    init(detail: CallRecordDetail) {
        switch detail.record.summaryState {
        case .unavailable:
            self = .hidden
        case .pending:
            self = .pending
        case .ready:
            self = .ready(detail.summary)
        case .failed:
            self = .failed(detail.summary)
        }
    }
}

struct CallDetailState: Equatable {
    var detail: CallRecordDetail?
    var timeline: [CallTimelineItem]
    var nextTimelineCursor: String?
    var timelineCollectionRevision: String?
    var syncStatus: CallHistorySyncStatus
    var errorCode: String?
    var errorMessage: String?
    var isLoadingMore: Bool

    var hasMoreTimeline: Bool { nextTimelineCursor != nil }
    var visibleTimeline: [CallTimelineItem] { timeline.filter { $0.kind != .unknown } }
    var isNormalNoAIContent: Bool {
        detail?.record.source == .remoteHandset
            && detail?.record.summaryState == .unavailable
            && detail?.record.hasTranscript == false
            && visibleTimeline.isEmpty
    }
}

@MainActor
final class CallHistoryModel: ObservableObject {
    private static let maxCachedRecords = 500
    private static let maxCachedDetails = 50
    private static let maxTimelineItems = 500

    @Published private(set) var records: [CallRecordItem] = []
    @Published private(set) var syncStatus: CallHistorySyncStatus = .idle
    @Published private(set) var errorCode: String?
    @Published private(set) var errorMessage: String?
    @Published private(set) var isRefreshing = false
    @Published private(set) var isLoadingMore = false
    @Published private(set) var collectionRevision: String?
    @Published private(set) var details: [String: CallDetailState] = [:]

    private let client: CallRecordContentClient
    private let store: CallHistoryCacheStoring
    private let deviceId: String
    private let clockMilliseconds: () -> Int64
    private let onUnauthorized: @MainActor () -> Void
    private var nextCursor: String?
    private var didLoadCache = false
    private var generation = 0
    private var loadingDetails = Set<String>()

    var hasMore: Bool { nextCursor != nil }

    init(
        client: CallRecordContentClient,
        store: CallHistoryCacheStoring,
        deviceId: String,
        clockMilliseconds: @escaping () -> Int64 = {
            Int64(Date().timeIntervalSince1970 * 1_000)
        },
        onUnauthorized: @escaping @MainActor () -> Void = {}
    ) {
        self.client = client
        self.store = store
        self.deviceId = deviceId
        self.clockMilliseconds = clockMilliseconds
        self.onUnauthorized = onUnauthorized
    }

    func detail(for callId: String) -> CallDetailState? { details[callId] }

    func loadCachedContent() {
        loadCacheIfNeeded()
    }

    func refresh() async {
        guard !isRefreshing, !isLoadingMore else { return }
        loadCacheIfNeeded()
        let requestGeneration = generation
        isRefreshing = true
        if records.isEmpty { syncStatus = .loading }
        defer { isRefreshing = false }
        do {
            let page = try await client.listCallRecords(limit: 25, cursor: nil)
            guard requestGeneration == generation else { return }
            let freshIds = Set(page.items.map(\.callId))
            records = Array(
                (page.items + records.filter { !freshIds.contains($0.callId) })
                    .prefix(Self.maxCachedRecords)
            )
            nextCursor = page.nextCursor
            collectionRevision = page.collectionRevision
            syncStatus = .live
            errorCode = nil
            errorMessage = nil
            saveCache()
        } catch {
            guard requestGeneration == generation else { return }
            handleListError(error)
        }
    }

    func loadMore() async {
        guard let cursor = nextCursor, !isRefreshing, !isLoadingMore else { return }
        let requestGeneration = generation
        isLoadingMore = true
        defer { isLoadingMore = false }
        do {
            let page = try await client.listCallRecords(limit: 25, cursor: cursor)
            guard requestGeneration == generation else { return }
            var indexes = Dictionary(uniqueKeysWithValues: records.enumerated().map { ($1.callId, $0) })
            for record in page.items {
                if let index = indexes[record.callId] {
                    records[index] = record
                } else {
                    indexes[record.callId] = records.count
                    records.append(record)
                }
            }
            records = Array(records.prefix(Self.maxCachedRecords))
            nextCursor = page.nextCursor
            collectionRevision = page.collectionRevision
            syncStatus = .live
            errorCode = nil
            errorMessage = nil
            saveCache()
        } catch {
            guard requestGeneration == generation else { return }
            handleListError(error)
        }
    }

    func refreshDetail(callId: String) async {
        guard !loadingDetails.contains(callId) else { return }
        loadCacheIfNeeded()
        let requestGeneration = generation
        loadingDetails.insert(callId)
        var state = details[callId] ?? Self.emptyDetailState
        if state.detail == nil { state.syncStatus = .loading }
        details[callId] = state
        defer { loadingDetails.remove(callId) }

        do {
            let detail = try await client.getCallRecord(callId: callId)
            guard requestGeneration == generation else { return }
            guard detail.record.callId == callId else { throw Self.invalidResponseError }
            state = details[callId] ?? Self.emptyDetailState
            state.detail = detail
            state.errorCode = nil
            state.errorMessage = nil
            details[callId] = state
            replaceListRecord(detail.record)
            saveCache()

            let timeline = try await client.listCallTimeline(callId: callId, limit: 50, cursor: nil)
            guard requestGeneration == generation else { return }
            state = details[callId] ?? Self.emptyDetailState
            state.timeline = Array(timeline.items.prefix(Self.maxTimelineItems))
            state.nextTimelineCursor = timeline.nextCursor
            state.timelineCollectionRevision = timeline.collectionRevision
            state.syncStatus = .live
            state.errorCode = nil
            state.errorMessage = nil
            details[callId] = state
            saveCache()
        } catch {
            guard requestGeneration == generation else { return }
            handleDetailError(error, callId: callId)
        }
    }

    func loadMoreTimeline(callId: String) async {
        guard var state = details[callId],
              let cursor = state.nextTimelineCursor,
              !state.isLoadingMore else { return }
        let requestGeneration = generation
        state.isLoadingMore = true
        details[callId] = state
        defer {
            if var latest = details[callId] {
                latest.isLoadingMore = false
                details[callId] = latest
            }
        }
        do {
            let page = try await client.listCallTimeline(callId: callId, limit: 50, cursor: cursor)
            guard requestGeneration == generation else { return }
            state = details[callId] ?? Self.emptyDetailState
            var indexes = Dictionary(uniqueKeysWithValues: state.timeline.enumerated().map { ($1.id, $0) })
            for item in page.items {
                if let index = indexes[item.id] {
                    state.timeline[index] = item
                } else {
                    indexes[item.id] = state.timeline.count
                    state.timeline.append(item)
                }
            }
            state.timeline = Array(state.timeline.prefix(Self.maxTimelineItems))
            state.nextTimelineCursor = page.nextCursor
            state.timelineCollectionRevision = page.collectionRevision
            state.syncStatus = .live
            state.errorCode = nil
            state.errorMessage = nil
            details[callId] = state
            saveCache()
        } catch {
            guard requestGeneration == generation else { return }
            handleDetailError(error, callId: callId)
        }
    }

    func clearLocalData() {
        generation += 1
        try? store.clear()
        records = []
        details = [:]
        nextCursor = nil
        collectionRevision = nil
        errorCode = nil
        errorMessage = nil
        syncStatus = .idle
        didLoadCache = true
    }

    private func loadCacheIfNeeded() {
        guard !didLoadCache else { return }
        didLoadCache = true
        guard let snapshot = try? store.load(deviceId: deviceId) else { return }
        records = Array(snapshot.records.prefix(Self.maxCachedRecords))
        collectionRevision = snapshot.collectionRevision
        details = snapshot.details.mapValues { cached in
            CallDetailState(
                detail: cached.detail,
                timeline: Array(cached.timeline.prefix(Self.maxTimelineItems)),
                nextTimelineCursor: cached.nextTimelineCursor,
                timelineCollectionRevision: cached.timelineCollectionRevision,
                syncStatus: .stale,
                errorCode: nil,
                errorMessage: nil,
                isLoadingMore: false
            )
        }
        if !records.isEmpty { syncStatus = .stale }
        if snapshot.records.count > Self.maxCachedRecords
            || snapshot.details.count > Self.maxCachedDetails
            || snapshot.details.values.contains(where: { $0.timeline.count > Self.maxTimelineItems }) {
            saveCache()
        }
    }

    private func replaceListRecord(_ record: CallRecordItem) {
        guard let index = records.firstIndex(where: { $0.callId == record.callId }) else { return }
        records[index] = record
    }

    private func handleListError(_ error: Error) {
        let hosted = error as? HostedCloudError
        if hosted?.code == "UNAUTHORIZED" {
            clearLocalData()
            errorCode = "UNAUTHORIZED"
            errorMessage = CallHistoryCopy.unauthorized
            syncStatus = .offline
            onUnauthorized()
            return
        }
        errorCode = hosted?.code
        errorMessage = Self.copy(for: hosted?.code)
        syncStatus = records.isEmpty ? .offline : .stale
    }

    private func handleDetailError(_ error: Error, callId: String) {
        let hosted = error as? HostedCloudError
        if hosted?.code == "UNAUTHORIZED" {
            clearLocalData()
            errorCode = "UNAUTHORIZED"
            errorMessage = CallHistoryCopy.unauthorized
            syncStatus = .offline
            onUnauthorized()
            return
        }
        var state = details[callId] ?? Self.emptyDetailState
        state.errorCode = hosted?.code
        state.errorMessage = Self.copy(for: hosted?.code)
        state.syncStatus = state.detail == nil ? .offline : .stale
        details[callId] = state
    }

    private func saveCache() {
        let allowedDetailIds = Set(records.prefix(Self.maxCachedDetails).map(\.callId))
        let cachedDetails = details.compactMapValues { state -> CachedCallDetail? in
            guard allowedDetailIds.contains(state.detail?.record.callId ?? ""),
                  let detail = state.detail else { return nil }
            return CachedCallDetail(
                detail: detail,
                timeline: Array(state.timeline.prefix(Self.maxTimelineItems)),
                nextTimelineCursor: state.nextTimelineCursor,
                timelineCollectionRevision: state.timelineCollectionRevision
            )
        }
        try? store.save(CallHistoryCacheSnapshot(
            deviceId: deviceId,
            records: Array(records.prefix(Self.maxCachedRecords)),
            collectionRevision: collectionRevision,
            details: cachedDetails,
            savedAt: clockMilliseconds()
        ))
    }

    private static func copy(for code: String?) -> String {
        switch code {
        case "PAYLOAD_TOO_LARGE": CallHistoryCopy.payloadTooLarge
        case "EDGE_OFFLINE", "TIMEOUT": CallHistoryCopy.edgeOffline
        case "FEATURE_DISABLED", "FORBIDDEN": CallHistoryCopy.featureDisabled
        case "UNAUTHORIZED": CallHistoryCopy.unauthorized
        default: CallHistoryCopy.unavailable
        }
    }

    private static let emptyDetailState = CallDetailState(
        detail: nil, timeline: [], nextTimelineCursor: nil,
        timelineCollectionRevision: nil, syncStatus: .idle,
        errorCode: nil, errorMessage: nil, isLoadingMore: false
    )
    private static let invalidResponseError = HostedCloudError(
        statusCode: 200,
        code: "INVALID_RESPONSE",
        message: L10n.text("calls.error.identity_mismatch")
    )
}

File: android/app/src/main/kotlin/ai/bondings/callpilot/content/CallHistoryModel.kt
package ai.bondings.callpilot.content

import ai.bondings.callpilot.protocol.CallRecordContentClient
import ai.bondings.callpilot.protocol.CallRecordDetail
import ai.bondings.callpilot.protocol.CallRecordItem
import ai.bondings.callpilot.protocol.CallSource
import ai.bondings.callpilot.protocol.CallSummaryState
import ai.bondings.callpilot.protocol.CallTimelineItem
import ai.bondings.callpilot.protocol.CallTimelineKind
import ai.bondings.callpilot.protocol.HostedCloudException
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext

enum class CallHistorySyncStatus { IDLE, LOADING, LIVE, STALE, OFFLINE }

enum class CallSummaryPresentation { HIDDEN, PENDING, READY, FAILED }

object CallHistoryCopy {
    const val PAYLOAD_TOO_LARGE = "这条通话内容过长，当前版本暂时无法显示"
    const val UNAVAILABLE = "暂时无法载入通话记录"
    const val EDGE_OFFLINE = "电脑端离线，正在显示本机缓存"
    const val FEATURE_DISABLED = "通话记录同步尚未启用"
    const val UNAUTHORIZED = "设备授权已失效，请重新配对"
}

data class CallDetailState(
    val detail: CallRecordDetail? = null,
    val timeline: List<CallTimelineItem> = emptyList(),
    val nextTimelineCursor: String? = null,
    val timelineCollectionRevision: String? = null,
    val syncStatus: CallHistorySyncStatus = CallHistorySyncStatus.IDLE,
    val errorCode: String? = null,
    val errorMessage: String? = null,
    val isLoadingMore: Boolean = false,
) {
    val hasMoreTimeline: Boolean get() = nextTimelineCursor != null
    val visibleTimeline: List<CallTimelineItem> get() = timeline.filter { it.kind != CallTimelineKind.UNKNOWN }
    val summaryPresentation: CallSummaryPresentation
        get() = when (detail?.record?.summaryState) {
            CallSummaryState.PENDING -> CallSummaryPresentation.PENDING
            CallSummaryState.READY -> CallSummaryPresentation.READY
            CallSummaryState.FAILED -> CallSummaryPresentation.FAILED
            CallSummaryState.UNAVAILABLE, null -> CallSummaryPresentation.HIDDEN
        }
    val isNormalNoAIContent: Boolean
        get() = detail?.record?.source == CallSource.REMOTE_HANDSET &&
            detail.record.summaryState == CallSummaryState.UNAVAILABLE &&
            !detail.record.hasTranscript && visibleTimeline.isEmpty()
}

data class CallHistoryState(
    val records: List<CallRecordItem> = emptyList(),
    val syncStatus: CallHistorySyncStatus = CallHistorySyncStatus.IDLE,
    val errorCode: String? = null,
    val errorMessage: String? = null,
    val isRefreshing: Boolean = false,
    val isLoadingMore: Boolean = false,
    val collectionRevision: String? = null,
    val hasMore: Boolean = false,
    val details: Map<String, CallDetailState> = emptyMap(),
)

class CallHistoryModel(
    private val client: CallRecordContentClient,
    private val store: CallHistoryCacheStoring,
    private val deviceId: String,
    private val clockMs: () -> Long = System::currentTimeMillis,
    private val onUnauthorized: () -> Unit = {},
) {
    private val mutableState = MutableStateFlow(CallHistoryState())
    val state: StateFlow<CallHistoryState> = mutableState.asStateFlow()

    private var nextCursor: String? = null
    private var cacheLoaded = false
    private val generation = AtomicInteger(0)
    private val loadingDetails = mutableSetOf<String>()

    suspend fun loadCachedContent() = loadCacheIfNeeded()

    suspend fun refresh() {
        if (state.value.isRefreshing || state.value.isLoadingMore) return
        loadCacheIfNeeded()
        val requestGeneration = generation.get()
        mutableState.value = state.value.copy(
            isRefreshing = true,
            syncStatus = if (state.value.records.isEmpty()) CallHistorySyncStatus.LOADING else state.value.syncStatus,
        )
        try {
            val page = client.listCallRecords(25, null)
            if (requestGeneration != generation.get()) return
            val freshIds = page.items.mapTo(mutableSetOf(), CallRecordItem::callId)
            val records = (page.items + state.value.records.filterNot { it.callId in freshIds }).take(MAX_RECORDS)
            nextCursor = page.nextCursor
            mutableState.value = state.value.copy(
                records = records,
                syncStatus = CallHistorySyncStatus.LIVE,
                errorCode = null,
                errorMessage = null,
                collectionRevision = page.collectionRevision,
                hasMore = nextCursor != null,
            )
            saveCache()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (requestGeneration == generation.get()) handleListError(error)
        } finally {
            if (requestGeneration == generation.get()) {
                mutableState.value = state.value.copy(isRefreshing = false)
            }
        }
    }

    suspend fun loadMore() {
        val cursor = nextCursor ?: return
        if (state.value.isRefreshing || state.value.isLoadingMore) return
        val requestGeneration = generation.get()
        mutableState.value = state.value.copy(isLoadingMore = true)
        try {
            val page = client.listCallRecords(25, cursor)
            if (requestGeneration != generation.get()) return
            val records = state.value.records.toMutableList()
            val indexes = records.mapIndexed { index, record -> record.callId to index }.toMap().toMutableMap()
            page.items.forEach { record ->
                val index = indexes[record.callId]
                if (index == null) {
                    indexes[record.callId] = records.size
                    records += record
                } else if (records[index].revision != record.revision) {
                    records[index] = record
                }
            }
            val bounded = records.take(MAX_RECORDS)
            nextCursor = page.nextCursor
            mutableState.value = state.value.copy(
                records = bounded,
                syncStatus = CallHistorySyncStatus.LIVE,
                errorCode = null,
                errorMessage = null,
                collectionRevision = page.collectionRevision,
                hasMore = nextCursor != null,
            )
            saveCache()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (requestGeneration == generation.get()) handleListError(error)
        } finally {
            if (requestGeneration == generation.get()) {
                mutableState.value = state.value.copy(isLoadingMore = false)
            }
        }
    }

    suspend fun refreshDetail(callId: String) {
        if (!loadingDetails.add(callId)) return
        loadCacheIfNeeded()
        val requestGeneration = generation.get()
        updateDetail(callId) { current ->
            if (current.detail == null) current.copy(syncStatus = CallHistorySyncStatus.LOADING) else current
        }
        try {
            val detail = client.getCallRecord(callId)
            if (requestGeneration != generation.get()) return
            if (detail.record.callId != callId) {
                throw HostedCloudException(200, "INVALID_RESPONSE", "通话详情标识不匹配")
            }
            updateDetail(callId) {
                it.copy(detail = detail, errorCode = null, errorMessage = null)
            }
            replaceListRecord(detail.record)
            saveCache()

            val timeline = client.listCallTimeline(callId, 50, null)
            if (requestGeneration != generation.get()) return
            updateDetail(callId) {
                it.copy(
                    timeline = timeline.items.take(MAX_TIMELINE_ITEMS),
                    nextTimelineCursor = timeline.nextCursor,
                    timelineCollectionRevision = timeline.collectionRevision,
                    syncStatus = CallHistorySyncStatus.LIVE,
                    errorCode = null,
                    errorMessage = null,
                )
            }
            saveCache()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (requestGeneration == generation.get()) handleDetailError(error, callId)
        } finally {
            loadingDetails.remove(callId)
        }
    }

    suspend fun loadMoreTimeline(callId: String) {
        val current = state.value.details[callId] ?: return
        val cursor = current.nextTimelineCursor ?: return
        if (current.isLoadingMore) return
        val requestGeneration = generation.get()
        updateDetail(callId) { it.copy(isLoadingMore = true) }
        try {
            val page = client.listCallTimeline(callId, 50, cursor)
            if (requestGeneration != generation.get()) return
            val timeline = (state.value.details[callId]?.timeline ?: emptyList()).toMutableList()
            val indexes = timeline.mapIndexed { index, item -> item.timelineItemId to index }.toMap().toMutableMap()
            page.items.forEach { item ->
                val index = indexes[item.timelineItemId]
                if (index == null) {
                    indexes[item.timelineItemId] = timeline.size
                    timeline += item
                } else {
                    timeline[index] = item
                }
            }
            updateDetail(callId) {
                it.copy(
                    timeline = timeline.take(MAX_TIMELINE_ITEMS),
                    nextTimelineCursor = page.nextCursor,
                    timelineCollectionRevision = page.collectionRevision,
                    syncStatus = CallHistorySyncStatus.LIVE,
                    errorCode = null,
                    errorMessage = null,
                )
            }
            saveCache()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (requestGeneration == generation.get()) handleDetailError(error, callId)
        } finally {
            if (requestGeneration == generation.get()) {
                updateDetail(callId) { it.copy(isLoadingMore = false) }
            }
        }
    }

    suspend fun clearLocalData() {
        generation.incrementAndGet()
        nextCursor = null
        cacheLoaded = true
        mutableState.value = CallHistoryState()
        withContext(Dispatchers.IO) { runCatching(store::clear) }
    }

    private suspend fun loadCacheIfNeeded() {
        if (cacheLoaded) return
        cacheLoaded = true
        val loadGeneration = generation.get()
        val snapshot = withContext(Dispatchers.IO) { runCatching { store.load(deviceId) }.getOrNull() }
            ?: return
        if (loadGeneration != generation.get()) return
        val records = snapshot.records.take(MAX_RECORDS)
        val allowedDetailIds = records.take(MAX_DETAILS).mapTo(mutableSetOf(), CallRecordItem::callId)
        val details = snapshot.details.mapNotNull { (callId, cached) ->
            if (callId !in allowedDetailIds) return@mapNotNull null
            callId to CallDetailState(
                detail = cached.detail,
                timeline = cached.timeline.take(MAX_TIMELINE_ITEMS),
                nextTimelineCursor = cached.nextTimelineCursor,
                timelineCollectionRevision = cached.timelineCollectionRevision,
                syncStatus = CallHistorySyncStatus.STALE,
            )
        }.toMap()
        mutableState.value = state.value.copy(
            records = records,
            syncStatus = if (records.isEmpty()) CallHistorySyncStatus.IDLE else CallHistorySyncStatus.STALE,
            collectionRevision = snapshot.collectionRevision,
            details = details,
        )
        if (
            snapshot.records.size > MAX_RECORDS || snapshot.details.size > MAX_DETAILS ||
            snapshot.details.values.any { it.timeline.size > MAX_TIMELINE_ITEMS }
        ) {
            saveCache()
        }
    }

    private fun replaceListRecord(record: CallRecordItem) {
        val index = state.value.records.indexOfFirst { it.callId == record.callId }
        if (index < 0) return
        val records = state.value.records.toMutableList().also { it[index] = record }
        mutableState.value = state.value.copy(records = records)
    }

    private suspend fun handleListError(error: Exception) {
        val code = (error as? HostedCloudException)?.errorCode
        if (code == "UNAUTHORIZED") {
            clearLocalData()
            mutableState.value = state.value.copy(
                syncStatus = CallHistorySyncStatus.OFFLINE,
                errorCode = code,
                errorMessage = CallHistoryCopy.UNAUTHORIZED,
            )
            onUnauthorized()
            return
        }
        mutableState.value = state.value.copy(
            syncStatus = if (state.value.records.isEmpty()) CallHistorySyncStatus.OFFLINE else CallHistorySyncStatus.STALE,
            errorCode = code,
            errorMessage = errorCopy(code),
        )
    }

    private suspend fun handleDetailError(error: Exception, callId: String) {
        val code = (error as? HostedCloudException)?.errorCode
        if (code == "UNAUTHORIZED") {
            handleListError(error)
            return
        }
        updateDetail(callId) {
            it.copy(
                syncStatus = if (it.detail == null) CallHistorySyncStatus.OFFLINE else CallHistorySyncStatus.STALE,
                errorCode = code,
                errorMessage = errorCopy(code),
            )
        }
    }

    private fun errorCopy(code: String?): String = when (code) {
        "PAYLOAD_TOO_LARGE" -> CallHistoryCopy.PAYLOAD_TOO_LARGE
        "EDGE_OFFLINE", "TIMEOUT" -> CallHistoryCopy.EDGE_OFFLINE
        "FEATURE_DISABLED", "FORBIDDEN" -> CallHistoryCopy.FEATURE_DISABLED
        "UNAUTHORIZED" -> CallHistoryCopy.UNAUTHORIZED
        else -> CallHistoryCopy.UNAVAILABLE
    }

    private fun updateDetail(callId: String, transform: (CallDetailState) -> CallDetailState) {
        val details = state.value.details.toMutableMap()
        details[callId] = transform(details[callId] ?: CallDetailState())
        mutableState.value = state.value.copy(details = details)
    }

    private suspend fun saveCache() {
        val saveGeneration = generation.get()
        val snapshotState = state.value
        val allowedDetailIds = snapshotState.records.take(MAX_DETAILS).mapTo(mutableSetOf(), CallRecordItem::callId)
        val details = snapshotState.details.mapNotNull { (callId, state) ->
            val detail = state.detail ?: return@mapNotNull null
            if (callId !in allowedDetailIds) return@mapNotNull null
            callId to CachedCallDetail(
                detail = detail,
                timeline = state.timeline.take(MAX_TIMELINE_ITEMS),
                nextTimelineCursor = state.nextTimelineCursor,
                timelineCollectionRevision = state.timelineCollectionRevision,
            )
        }.toMap()
        val snapshot = CallHistoryCacheSnapshot(
            deviceId = deviceId,
            records = snapshotState.records.take(MAX_RECORDS),
            collectionRevision = snapshotState.collectionRevision,
            details = details,
            savedAt = clockMs(),
        )
        withContext(Dispatchers.IO) {
            runCatching {
                store.save(snapshot)
                if (saveGeneration != generation.get()) store.clear()
            }
        }
    }

    private companion object {
        const val MAX_RECORDS = 500
        const val MAX_DETAILS = 50
        const val MAX_TIMELINE_ITEMS = 500
    }
}
